wire-ios-data-model/Source/ManagedObjectContext/Migration/118-119/FixDuplicateOneOnOneConversationsAction.swift

```diff
@@ -0,0 +1,128 @@
+//
+// Wire
+// Copyright (C) 2024 Wire Swiss GmbH
+//
+// This program is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU General Public License for more details.
+//
+// You should have received a copy of the GNU General Public License
+// along with this program. If not, see http://www.gnu.org/licenses/.
+//
+
+import Foundation
+
+final class FixDuplicateOneOnOneConversationsAction: CoreDataMigrationAction {
+
+    override func execute(in context: NSManagedObjectContext) throws {
+
+        // 1) fetch all conversations to fix
+        guard let fake1On1Conversations: [ZMConversation] = try? fakeOneOnOneConversations(context: context), !fake1On1Conversations.isEmpty else {
+            return
+        }
+
+        let fakeOneOnOnePerUser = conversationsPerUser(from: fake1On1Conversations)
+
+        for (user, conversations) in fakeOneOnOnePerUser {
+            guard conversations.count > 1 else {
+                // if only one conversation, nothing to fix
+                continue
+            }
+            var sortedConversations = conversations.sorted { lhs, rhs in
+                lhs.primaryKey < rhs.primaryKey
+            }
+
+            // 2) set 1:1 conversation
+            let theOneOnOneConversation = sortedConversations.removeFirst()
+            user.oneOnOneConversation = theOneOnOneConversation
+
+            WireLogger.localStorage.debug("Fixing oneOnOne conversation", attributes: [.conversationId: theOneOnOneConversation.remoteIdentifier.safeForLoggingDescription])
+
+            // 3) select all messages from other conversations
+            for otherConversation in sortedConversations {
+                moveMessages(from: otherConversation, to: theOneOnOneConversation)
+            }
+        }
+    }
+
+    private func conversationsPerUser(from conversations: [ZMConversation]) -> [ZMUser: [ZMConversation]] {
+        var fakeOnOnOnePerUser = [ZMUser: [ZMConversation]]()
+        for conversation in conversations {
+            if let otherUser = conversation.localParticipantsExcludingSelf.first {
+                if fakeOnOnOnePerUser[otherUser] == nil {
+                    fakeOnOnOnePerUser[otherUser] = [conversation]
+                } else {
+                    fakeOnOnOnePerUser[otherUser]?.append(conversation)
+                }
+
+            }
+        }
+        return fakeOnOnOnePerUser
+    }
+
+    private func moveMessages(from otherConversation: ZMConversation,
+                              to theOneOnOneConversation: ZMConversation) {
+        otherConversation.allMessages.forEach { message in
```

Comment:
Optional: I don't think the specifics of the content in the method are tested... eg. visible vs hidden, draft properties. Consider testing this if it is not too much trouble.